#10514 - FeatureEditor filter by geometric area

Gaetanbrl opened the pull request:
## Description

This improvement will change the featureEditor plugin to allow to limit attributes table features loading by a geometric area (WKT or GeoJSON) for non ADMIN users.

This improvement don't use GeoFence and use a simple geometry.

>https://groups.google.com/g/mapstore-developers/c/6KhmNTJzpDc/m/12J-2t9BAAAJ

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue

**What is the current behavior?**

User can filter attribute table by custom filter or viewport.

#10514 

**What is the new behavior?**

If configured, attribute table keep natives filters and use an additional geometric filter that can be create from URL (e.g geOrchestra restricted area) or from RAW (GEOJSON or WKT).

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

Gaetanbrl committed:
Config, button,  reux, filter for restricted area

Gaetanbrl committed:
keep existing filters without viewport

Gaetanbrl committed:
Change restricted area setter

restricted area documentation

Lint and clean

tdipisa commented:
Hi @Gaetanbrl thank you for contributing. We will review asap.

offtherailz commented on web/client/selectors/featuregrid.js:
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
+ * Create spatialField filters array.
+ * Contains filters from viewportFilter, restrictedArea, exsting WFS filter
+ */
+export const additionnalGridFilters = (state) => {
```
Using this approach makes the sync with map not working. While it still makes sense with viewport filter (because the map already shows the features in map), here this makes things a little unconsistent, and it could be practical for the user to see only the allowed rows on the map. What do you think?

A more functional approach is to add / remove the filter to the `query` object, as well as the other filters, like quick filter, do. 
![image](https://github.com/user-attachments/assets/1e46db98-300b-47b3-a22d-35fee67102e0)

The  `filters` [section](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)  recently added to the filter may help you to manage it in a easier way.

Basically you can add to the filter a section filters: that can contain and use cql filter to apply some additional logic. 
You can identify these filters by id, adding your own property, to manage them (e.g. adding/removing).

Gaetanbrl replied:
> it could be practical for the user to see only the allowed rows on the map. What do you think?

Exactly, this is the expected behavior

> Using this approach makes the sync with map not working. 

Not really sur to understand. additionnalGridFilters methods is not dependent on the screen filter but but also uses it if activate. Maybe my code is wrong with I wanted to do.

> Basically you can add to the filter a section filters: that can contain and use cql filter to apply some additional logic.

I will change the code to use filters section as you expect.

Thanks for this proposals @offtherailz .

MV88 replied:
is this done?

Gaetanbrl replied:
Not yet.

Gaetanbrl replied:
I will use this filter instead of spatialFilter : 

```
    {
            "format": "cql",
            "version": "1.0.0",
            "body": `${restrictedAreaOperatorSelector(state)}(${attribute}, ${asWKT})`,
            "id": `[${uniqueId("cql_restricted_area_")}]`,
            "restrictedArea": true
    };
```

offtherailz commented on web/client/plugins/FeatureEditor.jsx:
```diff
@@ -79,6 +79,9 @@ import {isViewportFilterActive} from "../selectors/featuregrid";
   * @prop {array} cfg.showFilterByViewportTool Show button to toggle filter by viewport in toolbar.
   * @prop {object} cfg.dateFormats Allows to specify custom date formats ( in [ISO_8601](https://en.wikipedia.org/wiki/ISO_8601)  format) to use to display dates in the table. `date` `date-time` and `time` are the supported entries for the date format. Example:
   * @prop {boolean} cfg.showPopoverSync default false. Hide the popup of map sync if false, shows the popup of map sync if true
+  * @prop {string} cfg.restrictedArea.url Geometry definition as WKT or GeoJSON loaded from URL or path.
+  * @prop {string} cfg.restrictedArea.raw Geometry definition as WKT or GeoJSON.
+  * @prop {string} cfg.restrictedArea.operator Spatial operation to performed between features and the given geometry.
```
```suggestion
  * @prop {object} cfg.restrictedArea object containing settings for restricted area. If present, it restricts the editing area to the given geometry. It requires at least `url` or `raw` to be defined.
  * @prop {string} cfg.restrictedArea.raw Geometry definition as WKT or GeoJSON. This attribute allows to define the geometry directly in the configuration.
  * @prop {string} cfg.restrictedArea.url Geometry definition as WKT or GeoJSON loaded from URL or path. If present, this wins over the raw geometry configuration. By default, the filter will use `CONTAINS` if not defined.
```

MV88 replied:
is this done?

Gaetanbrl replied:
Not yet.

Gaetanbrl replied:
Modified and pushed. Thanks to you.

offtherailz reviewed (changes requested):
Thank you for the PR.  In this review I reported all the things that I found that may look unconsistent or problematic.

- [ ] see my comments inline
- [ ] The original proposal and issue was regarding to restrict editing, but in this implementation, also the view mode encounter this filter. Also this makes the things a little unclear. Maybe the filter should be applied only in edit mode?
- [ ] I found a bug that I didn't investigated, so I write about it here, instead of inline. With this tool active, I see an error of "max stack exceed" as well as I try to apply a quick filter.
    https://github.com/user-attachments/assets/0193f17b-bfc2-41a2-a152-184e9482c445
    I didn't debug so I don't know the exact reason. Anyway using `spatialField` can problematic, because too related to the advanced filter form . I'd suggest to use "[filters](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)" instead, as reported in another inline comment.

Gaetanbrl commented:
@offtherailz Thank you for this detailed review.
I'll think about functional proposals with and try to plan technical changes.

Gaetanbrl commented:
The source branch is now rebase / align with master (https://github.com/geosolutions-it/MapStore2/commit/8e2d4433427ce5b7cf7628c8d51f724f29d2bf18)

Gaetanbrl commented on web/client/epics/featuregrid.js:
```diff
@@ -1283,3 +1298,34 @@ export const resetViewportFilter = (action$, store) =>
+export const requestRestrictedArea = (action$, store) =>
+    action$.ofType(OPEN_FEATURE_GRID, LOGIN_SUCCESS)
+        .filter(() => {
+            return !isAdminUserSelector(store.getState())
```
Indeed, we have think avout authenticated users (to works with geOrchestra restricted org area) first and forget anonymous users.

To apply this restriction, we needs to confirm if restricted area is apply in edit mode only or not as explain in global review comment.

Gaetanbrl replied:
> I think we should at least apply the same restrictedArea to the anonymous users?

After discussion, we will also apply the restriction to anonymous users to be compliant with MapStore2 behavior.

MV88 replied:
is this done?

Gaetanbrl replied:
Not yet.

Gaetanbrl replied:
Via https://github.com/geosolutions-it/MapStore2/pull/10515/commits/be5120d0027b335d7f945ce3706d813f31c78b6c commit , everyone will be restricted by default excepted admins.

@offtherailz is this enough ?

Gaetanbrl commented on web/client/components/data/featuregrid/toolbars/Toolbar.jsx:
```diff
@@ -35,7 +35,23 @@ const standardButtons = {
         onClick={events.switchEditMode}
-        glyph="pencil"/>),
+        glyph="pencil" />),
+    isRestrictedByArea: ({ restrictedArea }) => {
+        return (<Button
+            id="fg-isRestrictedByArea-button"
+            keyProp="fg-restrictedarea-button"
+            className="square-button-md"
+            bsStyle="warning"
+            tooltipId="featuregrid.toolbar.restrictedByArea"
+            disabled
+            style={isEmpty(restrictedArea) ? {
+                width: 0,
+                padding: 0,
+                borderWidth: 0
+            } : {}}>
+            <Glyphicon glyph="1-point-dashed" />
+        </Button>);
```
Thanks. This change will be made.

MV88 replied:
is this suggestion applied?

Gaetanbrl replied:
Not yet.

Gaetanbrl replied:
About this change : 

> replace Button by TButton

Modification made and pushed. Thanks.

Gaetanbrl commented:
> Thank you for the PR. In this review I reported all the things that I found that may look unconsistent or problematic.
> 
> * [ ]  see my comments inline
> * [ ]  The original proposal and issue was regarding to restrict editing, but in this implementation, also the view mode encounter this filter. Also this makes the things a little unclear. Maybe the filter should be applied only in edit mode?
> * [ ]  I found a bug that I didn't investigated, so I write about it here, instead of inline. With this tool active, I see an error of "max stack exceed" as well as I try to apply a quick filter.
>   https://github.com/user-attachments/assets/0193f17b-bfc2-41a2-a152-184e9482c445
>   I didn't debug so I don't know the exact reason. Anyway using `spatialField` can problematic, because too related to the advanced filter form . I'd suggest to use "[filters](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)" instead, as reported in another inline comment.

After some discussions, we will make the expected changes (see comments).

> Maybe the filter should be applied only in edit mode?

About this point, the natif request is to limit view by an area in edit mode and read mode.

The natif use case is to limit view and edition according to geOrchestra restricted area (this area is applied to the user's organization via geOrchestra console app).

So, we propose to add a configuration to manage 2 options :

- allow to restrict edition only (to take your request into account)
- allow to restrict edition and view


> I see an error of "max stack exceed" as well as I try to apply a quick filter.

Needs investigation. I've seen this error before with CQL filters and geoServer limits (cross layer filter limit ?).

MV88 commented on web/client/selectors/featuregrid.js:
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
         return viewportFilterIsActive && viewportFilterIsSupported ? {
             spatialField: [
-                ...existingFilter,
+                // avoid restricted area filter dupplication
```
```suggestion
                // avoid restricted area filter duplication
```

Gaetanbrl replied:
Modified and pushed. Thanks to you.

MV88 commented on web/client/selectors/featuregrid.js:
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
+    viewportFilter,
+    projectionSelector,
+    describeSelector,
+    state => restrictedAreaOperatorSelector(state),
```
```suggestion
   restrictedAreaOperatorSelector,
```

MV88 commented on web/client/plugins/FeatureEditor.jsx:
```diff
@@ -175,7 +183,9 @@ const EditorPlugin = connect(
                 editingAllowedGroups: this.props.editingAllowedGroups,
-                maxStoredPages: this.props.maxStoredPages
+                maxStoredPages: this.props.maxStoredPages,
+                restrictedAreaUrl: this.props.restrictedAreaUrl,
```
this is not present in config, only restrictedArea with inside three params:
- raw
- operator
- url

MV88 commented on web/client/reducers/featuregrid.js:
```diff
@@ -156,7 +157,9 @@ function featuregrid(state = emptyResultsState, action) {
             virtualScroll: !!action.options.virtualScroll,
-            maxStoredPages: action.options.maxStoredPages || 5
+            maxStoredPages: action.options.maxStoredPages || 5,
+            restrictedAreaUrl: action.options.restrictedAreaUrl || "",
```
this is not needed as url is already inside restrictedArea

MV88 commented on web/client/reducers/featuregrid.js:
```diff
@@ -441,6 +444,9 @@ function featuregrid(state = emptyResultsState, action) {
     case MAP_CONFIG_LOADED: {
         return {...state, ...get(action, 'config.featureGrid', {})};
     }
+    case SET_RESTRICTED_AREA: {
```
add unit test for this

MV88 commented on web/client/reducers/featuregrid.js:
```diff
@@ -156,7 +157,9 @@ function featuregrid(state = emptyResultsState, action) {
-            maxStoredPages: action.options.maxStoredPages || 5
+            maxStoredPages: action.options.maxStoredPages || 5,
+            restrictedAreaUrl: action.options.restrictedAreaUrl || "",
+            restrictedArea: action.options.restrictedArea || {}
```
add unit test for this

MV88 commented on web/client/utils/FeatureGridUtils.js:
```diff
@@ -392,3 +394,38 @@ export const supportsFeatureEditing = (layer) => includes(supportedEditLayerType
  */
 export const areLayerFeaturesEditable = (layer) =>  !layer?.disableFeaturesEditing && supportsFeatureEditing(layer);
+
+export const isWKT = (wktString) => {
```
is this done?

Gaetanbrl replied:
Not yet.

Gaetanbrl replied:
After some code seek,
I will use [toGeoJson (from ogc/WKS utils)](https://github.com/geosolutions-it/MapStore2/blob/02520e5c0f3d97fb5df1a24c6b51802e5ba61ceb/web/client/utils/ogc/WKT/toGeoJSON.js#L147). Seems correct to detect if a string return a correct geometry.

MV88 reviewed (changes requested):
Hi @Gaetanbrl I resumed the review recently and i think you are still working on it right?
if so consider this as a partial review, since changes still have to be submitted yet

all of the following must have unit tests covering changes
- selectors 
- reducers 
- actions 
- epics

please checkout also inline comments

also if you can go through the various conversation and add a comment to say which can be resolved or not.

Thanks

Gaetanbrl commented:
@MV88 thank you for this review and your precious time (which we all miss).

> I resumed the review recently and i think you are still working on it right?

Yes, but last comment ([here](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2500206650)) waiting for some feedback about [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2443828702) and [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2500203185).
Especially on my counter proposal (see comment).

So, I didn't move forward, having no additional interaction.

I've been away since your last comment, I'll get back now to work on this contribution to change technical point but I always need a feedback about functional elements.

Gaetanbrl committed:
requested > changes around doc and code optimization

Gaetanbrl commented on web/client/selectors/featuregrid.js:
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
     return (canEdit || isAllowed) && !isCesium(state);
 };
+
+export const restrictedAreaSrcSelector = state => get(state, "featuregrid.restrictedArea");
```
Modified and pushed. Thanks to you.

Gaetanbrl committed:
use native ogc WKT tools

Gaetanbrl committed:
optimize code

Gaetanbrl committed:
restriction affect anonymous and logged user

Gaetanbrl committed:
remove unused

Gaetanbrl committed:
restriction affect anonymous and logged user

Gaetanbrl committed:
remove unused

Gaetanbrl commented:
@MV88 you answer "Is this done ?" about some remaining returns :

- [here](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1925726388)
- [here](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1925723669)

It would be really nice if you could give me some feedback on this

- https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2443828702
- especially about "How to use filters correctly ?", because I understand the documentation and how it works, but how can I use it properly in mapstore code ?


Thank you very much for these clarifications (@offtherailz).

Gaetanbrl committed:
Add i18n tooltip

Gaetanbrl committed:
Fix maximum call size error, use CQL with Filters

Gaetanbrl committed:
Only restrict by area on EDIT mode

Gaetanbrl committed:
remove comment

Gaetanbrl committed:
Fix cql filter srid syntax to sync wms and attribute table

Gaetanbrl commented:
Here is a summary of the covered items.

## Changes Following Reviews

- Fixed the *maximum call stack* error.
- Applied **restricted area** logic only in **EDIT** mode.
- **Restricted area** is **never** applied for **ADMIN** users.
- **Restricted area** is applied only for **USER** and **ANONYMOUS** profiles.
- Added option to enable synchronization between the attribute table and the map (see [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1808831879)).
- Various technical feedback items addressed (@MV88).

## Outstanding Points

- Unit tests have not been implemented yet.
- There is an issue when synchronization is enabled and then the user switches to edit mode — the CQL syntax becomes invalid  

 This last error seems related to `srid` and CQL syntax. Refer to the [google group dev](https://groups.google.com/g/mapstore-developers/c/lJFzee_0oBI) ticket for more details.

Gaetanbrl commented:
Same as #10524.

I can look into updating the PR, but I’m not certain it will still be sufficient...

tdipisa commented:
Hi @Gaetanbrl as you told me during our exchange last week, I'm going to close this.
